### ISSUE TITLE
[RFC]: Single-Node D2D Transfer - CUDA IPC Connector

### ISSUE THREAD

**Issue by ZhengWG** [vllm-project/vllm-omni#2379]
### Motivation.

In single-node multi-GPU deployments (e.g., Thinker on GPU0, Talker on GPU1), inter-stage tensor transfer is bottlenecked by two layers of CPU involvement:

1. **Payload processor layer:** `thinker2talker_async_chunk` calls `.detach().cpu()` on all tensors before passing them to the connector, forcing a D2H copy unconditionally.
2. **Connector layer:** Neither `SharedMemoryConnector` (CPU `/dev/shm`) nor `MooncakeTransferEngineConnector` (designed for cross-node RDMA) provides a GPU-direct transfer path for single-node use.

The resulting data flow is: GPU0 -> D2H -> serialize -> transport -> deserialize -> H2D -> GPU1, achieving only ~0.7 GB/s. CUDA IPC D2D via NVLink achieves ~350 GB/s on H20 -- over **two orders of magnitude faster**.Beyond raw transfer latency, the CPU bounce path introduces **pipeline bubble overhead** that is difficult to isolate in micro-benchmarks but accumulates in production.

### Proposed Change.

**Overview of Design:** 

Split-plane architecture:

- **Data Plane:** GPU tensors stay on sender GPU. CUDA IPC handles (64 bytes each) are passed to the receiver, which opens them and performs D2D copy via NVLink.
- **Control Plane:** IPC handles + small CPU data serialized via `/dev/shm` (< 1 KB per chunk).

Key decisions:

1. Use PyTorch native `torch.multiprocessing` IPC reduction (wraps `cudaIpcGetMemHandle`/`cudaIpcOpenMemHandle`).
2. ACK-based lifecycle: sender holds tensor refs until receiver completes D2D copy and sends ACK.
3. `supports_gpu_tensor` class attribute signals payload processor to skip `.cpu()`.

<img width="854" height="262" alt="Image" src="https://github.com/user-attachments/assets/4e593940-3c8c-4d45-98b0-6502980b2aa9" />

For more details, can refer to: [vLLM-Omni CUDA IPC D2D Connector Design Doc](https://docs.google.com/document/d/1BFNx6UAYOJtMsz-q-ayPLIVzhYmXKFflUn_HJsD0rq8/edit?usp=sharing)

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by hsliuustc0106** [comment]
@amy-why-3459 @natureofnature

**Comment by amy-why-3459** [comment]
For the D2D connector, do we need to apply for GPU memory in advance?

**Comment by ZhengWG** [comment]
> For the D2D connector, do we need to apply for GPU memory in advance?

No, CUDA IPC has no memory registration step, unlike MooncakeTransferEngineConnector.
For Sender, it just gets an IPC handle to the existing model output tensor. For Receiver, it can allocate via PyTorch's CUDA caching allocator.
If allocator contention becomes a bottleneck under high concurrency, we can add an optional pre-allocated pool later. Start simple.

**Comment by npuichigo** [comment]
Any update on this?

**Comment by ZhengWG** [comment]
> Any update on this?

Core feature completed in this related PR — targeting release this week.
